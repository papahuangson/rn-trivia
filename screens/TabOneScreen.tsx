import React, {useState, useEffect} from 'react';
import axios from 'axios';
import { StyleSheet, ScrollView, TouchableOpacity, ActivityIndicator } from 'react-native';

import { QuestionItem } from '../components/QuestionItem';
import { Text } from '../components/Themed';
import AsyncStorage from '@react-native-async-storage/async-storage';

export default function TabOneScreen() {
  let scrollView: any;

  const [loading, setLoading] = useState<boolean>(false);
  const [questions, setQuestions] = useState<any[]>([]);
  const [gameInProgress, setGameInProgress] = useState<boolean>(false);
  const [answers, updateAnswers] = useState<boolean[]>([]);

  const fetchData: () => Promise<void> = async () => {
    const result: any = await axios('https://opentdb.com/api.php?amount=5');
    const items = result.data.results;
    for (let i = 0; i < items.length; i++) {
      items[i].answers = randomizeAnswers([...items[i].incorrect_answers, items[i].correct_answer]);
    }
    setQuestions(items);
    setLoading(false);
  };

  const randomizeAnswers = (answers: any[]) => {
    //https://github.com/coolaj86/knuth-shuffle
    var currentIndex = answers.length,  randomIndex;

    // While there remain elements to shuffle...
    while (0 !== currentIndex) {

      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex--;

      // And swap it with the current element.
      [answers[currentIndex], answers[randomIndex]] = [
        answers[randomIndex], answers[currentIndex]];
    }

    return answers;
  }

  const onAnswer = (isCorrect: boolean) => {
    updateAnswers(arr => [...arr, isCorrect]);
    if (answers.length + 1 === 5) {
      updateLocalStats();
      scrollView.scrollToEnd();
    }
  }

  const updateLocalStats = async () => {
    try {
      const numberOfGames = await AsyncStorage.getItem('@numberOfGames');
      const correctAnswers = await AsyncStorage.getItem('@correctAnswers');
      await AsyncStorage.setItem('@numberOfGames', (numberOfGames ? Number(numberOfGames) + 1 : 1).toString());
      await AsyncStorage.setItem('@correctAnswers', (correctAnswers ? Number(correctAnswers) + answers.filter(Boolean).length : answers.filter(Boolean).length).toString());      
    } catch (e) {
      console.log(e);
    }
  }

  return (
    <ScrollView
      ref={ref => scrollView = ref}
      style={{paddingVertical: 20, flex: 1}}
      contentContainerStyle={styles.container}>
        {!loading ? (
          <>
            {questions.map((question: any, index: number) => <QuestionItem question={question} key={index} onAnswer={(isCorrect: boolean) => onAnswer(isCorrect)} />)}
            {answers.length === 5 && (
              <Text style={styles.correctAnswersText}>{answers.filter(Boolean).length}/5 Correct</Text>
            )}
            <TouchableOpacity testID="startGameButton" onPress={() => {
              updateAnswers([]);
              setQuestions([]);
              setGameInProgress(true);
              fetchData();
              setLoading(true);
            }}>
              <Text style={styles.startGameButtonText}>{answers.length === 5 ? 'Play again' : 'Start a new round'}</Text>
            </TouchableOpacity>
          </>
        ) : (
          <ActivityIndicator size="large" />
        )}
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  startGameButtonText: {
    fontSize: 20,
    padding: 20,
    backgroundColor: 'red',
    color: 'white',
  },
  correctAnswersText: {
    fontSize: 24,
    margin: 20,
    marginTop: 200,
  },
});
