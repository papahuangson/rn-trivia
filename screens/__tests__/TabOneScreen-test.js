import * as React from 'react';
import renderer from 'react-test-renderer';
import TabOneScreen from '../TabOneScreen';

let tree;
it(`renders correctly`, () => {
  tree = renderer.create(<TabOneScreen />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should have a button on load', () => {
  const button = tree.findWhere(t => t.prop('testID') === 'startGameButton');
})