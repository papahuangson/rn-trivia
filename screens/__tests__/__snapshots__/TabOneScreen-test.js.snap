// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
  style={
    Object {
      "flex": 1,
      "paddingVertical": 20,
    }
  }
>
  <View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="startGameButton"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            Object {
              "backgroundColor": "red",
              "color": "white",
              "fontSize": 20,
              "padding": 20,
            },
          ]
        }
      >
        Start a new round
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
